react/src/lib/components/DeckGLMap/layers/wells/wellsLayer.ts

```diff
@@ -348,3 +353,159 @@ function getLogProperty(
     if (log_value) return { name: prop_name, value: log_value };
     else return null;
 }
+
+function CatmullRom(
+    P0: Position3D,
+    P1: Position3D,
+    P2: Position3D,
+    P3: Position3D,
+    t: number
+): Position3D {
+    const alpha = 0.5;
+    const tt = t * t;
+    const ttt = t * t * t;
+
+    // disable eslint for some lines due to readability.
+    const dist_p0_p1 = Math.sqrt((P1[0]-P0[0])*(P1[0]-P0[0]) + (P1[1]-P0[1])*(P1[1]-P0[1]) + (P1[2]-P0[2])*(P1[2]-P0[2]) ); // eslint-disable-line
+    const dist_p1_p2 = Math.sqrt((P1[0]-P2[0])*(P1[0]-P2[0]) + (P1[1]-P2[1])*(P1[1]-P2[1]) + (P1[2]-P2[2])*(P1[2]-P2[2]) ); // eslint-disable-line
+    const dist_p2_p3 = Math.sqrt((P3[0]-P2[0])*(P3[0]-P2[0]) + (P3[1]-P2[1])*(P3[1]-P2[1]) + (P3[2]-P2[2])*(P3[2]-P2[2]) ); // eslint-disable-line
+
+    const t01 = Math.pow(dist_p0_p1, alpha);
+    const t12 = Math.pow(dist_p1_p2, alpha);
+    const t23 = Math.pow(dist_p2_p3, alpha);
+
+    const m1_x = (P2[0] - P1[0] + t12 * ((P1[0] - P0[0]) / t01 - (P2[0] - P0[0]) / (t01 + t12))); // eslint-disable-line
+    const m1_y = (P2[1] - P1[1] + t12 * ((P1[1] - P0[1]) / t01 - (P2[1] - P0[1]) / (t01 + t12))); // eslint-disable-line
+    const m1_z = (P2[2] - P1[2] + t12 * ((P1[2] - P0[2]) / t01 - (P2[2] - P0[2]) / (t01 + t12))); // eslint-disable-line
+
+    const m2_x = (P2[0] - P1[0] + t12 * ((P3[0] - P2[0]) / t23 - (P3[0] - P1[0]) / (t12 + t23))); // eslint-disable-line
+    const m2_y = (P2[1] - P1[1] + t12 * ((P3[1] - P2[1]) / t23 - (P3[1] - P1[1]) / (t12 + t23))); // eslint-disable-line
+    const m2_z = (P2[2] - P1[2] + t12 * ((P3[2] - P2[2]) / t23 - (P3[2] - P1[2]) / (t12 + t23))); // eslint-disable-line
+
+    const a_x = 2 * (P1[0] - P2[0]) + m1_x + m2_x;
+    const a_y = 2 * (P1[1] - P2[1]) + m1_y + m2_y;
+    const a_z = 2 * (P1[2] - P2[2]) + m1_z + m2_z;
+
+    const b_x = -3 * (P1[0] - P2[0]) - m1_x - m1_x - m2_x;
+    const b_y = -3 * (P1[1] - P2[1]) - m1_y - m1_y - m2_y;
+    const b_z = -3 * (P1[2] - P2[2]) - m1_z - m1_z - m2_z;
+
+    const c_x = m1_x;
+    const c_y = m1_y;
+    const c_z = m1_z;
+
+    const d_x = P1[0];
+    const d_y = P1[1];
+    const d_z = P1[2];
+
+    const x = a_x * ttt + b_x * tt + c_x * t + d_x;
+    const y = a_y * ttt + b_y * tt + c_y * t + d_y;
+    const z = a_z * ttt + b_z * tt + c_z * t + d_z;
+
+    return [x, y, z] as Position3D;
+}
+
+function splineRefine(data, refine: boolean) {
+    const ts = refine ? [0.2, 0.4, 0.6, 0.8] : [];
+
+    if (data["features"] === undefined) {
+        return;
+    }
+
+    const no_wells = data["features"].length;
+    for (let well_no = 0; well_no < no_wells; well_no++) {
+        const mds = data["features"][well_no]["properties"]["md"];
+
+        const coords = data["features"][well_no]["geometry"]["geometries"][1]["coordinates"]; // eslint-disable-line
```

Comment:
Is it possible that any of these fields are not there for some trajectories? Or maybe there aren't at least 2 geometries, so `...["geometries"][1]["blabla"]` will throw an error? If the format guaranties that you have all those fields and number of geometries, then there's no action to be taken here, otherwise you probably need to add some checks.

Reply:
The dead code will be fixed in the next pull request. I will also then add a test to ensure data["features"] is defined. As for the other fields they are guaranteed by the format I think. The number "1" there for instance refers to the trajectory itself and "0" is the well head both of which has to be there

Reply:
The well heads and trajectories together might not work in the future when adding other well markers (where the well intersects surfaces, for example), but that's a discussion for another time :)